File: app/js/wikidata.js
/**
 * Looking up films the bundled library doesn't know, on Wikidata.
 *
 * No key and no account: Wikidata and Wikipedia allow browser requests from
 * anywhere. Only a film's title and year are sent - never a rating, and
 * nothing about the person. Results are cached in the browser, so each film is
 * only ever looked up once.
 *
 * Mirrors festrec_eval/wikidata.py and festrec_eval/genres.py: a film looked up
 * here must come out in the same shape, with genres in the same vocabulary, as
 * the ones the model was trained on. Change one, change both.
 */

const WIKIDATA = 'https://www.wikidata.org/w/api.php';
const WIKIPEDIA = 'https://en.wikipedia.org/w/api.php';
const CAST_DEPTH = 6;
const OVERVIEW_CHARS = 700;

const FILM_TYPES = new Set([
  'Q11424', 'Q24869', 'Q29168811', 'Q202866', 'Q93204', 'Q226730', 'Q506240',
]);
const TYPE_GENRES = {
  Q29168811: 'animation', Q202866: 'animation', Q17517379: 'animation',
  Q93204: 'documentary', Q24865: 'documentary',
};

export const GENRES = [
  'action', 'adventure', 'animation', 'comedy', 'crime', 'documentary', 'drama',
  'family', 'fantasy', 'history', 'horror', 'music', 'mystery', 'romance',
  'science fiction', 'thriller', 'war', 'western',
];

const WORDS = {
  action: 'action', adventure: 'adventure', animation: 'animation', animated: 'animation',
  'stop-motion': 'animation', anime: 'animation', comedy: 'comedy', 'dark comedy': 'comedy',
  'black comedy': 'comedy', comedic: 'comedy', crime: 'crime', documentary: 'documentary',
  doc: 'documentary', drama: 'drama', family: 'family', fantasy: 'fantasy',
  'dark fantasy': 'fantasy', history: 'history', historical: 'history', horror: 'horror',
  'creature horror': 'horror', 'body horror': 'horror', 'folk horror': 'horror',
  slasher: 'horror', supernatural: 'horror', music: 'music', musical: 'music',
  mystery: 'mystery', romance: 'romance', romantic: 'romance', 'rom-com': 'romance',
  'science fiction': 'science fiction', 'sci-fi': 'science fiction',
  scifi: 'science fiction', thriller: 'thriller', war: 'war', western: 'western',
  'neo-western': 'western', satire: 'comedy', parody: 'comedy', splatter: 'horror',
  gory: 'horror', ghost: 'horror', giallo: 'horror', creature: 'horror', zombie: 'horror',
  vampire: 'horror', psychodrama: 'drama', melodrama: 'drama', noir: 'thriller',
  'neo-noir': 'thriller', heist: 'crime', gangster: 'crime', spy: 'thriller',
  espionage: 'thriller', survival: 'thriller', kaiju: 'science fiction',
  cyberpunk: 'science fiction', dystopian: 'science fiction', space: 'science fiction',
  swordplay: 'action', martial: 'action', wuxia: 'action', revenge: 'thriller',
  erotic: 'romance', concert: 'music', biographical: 'history', biopic: 'history',
  docudrama: 'documentary', superhero: 'action', disaster: 'action', detective: 'mystery',
  whodunit: 'mystery', 'coming-of-age': 'drama', sports: 'drama', legal: 'drama',
  teen: 'drama', christmas: 'family', "children's": 'family', 'fairy tale': 'fantasy',
  'sword and sorcery': 'fantasy',
};

const COMPOUNDS = {
  'romantic comedy': ['romance', 'comedy'], 'rom-com': ['romance', 'comedy'],
  'comedy-drama': ['comedy', 'drama'], 'comedy drama': ['comedy', 'drama'],
  dramedy: ['comedy', 'drama'], 'horror comedy': ['horror', 'comedy'],
  'comedy horror': ['horror', 'comedy'], 'musical comedy': ['music', 'comedy'],
  'science fiction comedy': ['science fiction', 'comedy'],
  'action comedy': ['action', 'comedy'],
};

const MEDIUM = /\b(feature |short |television |tv )?(film|movie|cinema)s?\b/g;

export function cleanGenre(phrase) {
  return String(phrase).toLowerCase().replace(MEDIUM, '').replace(/\s+/g, ' ')
    .replace(/^[\s,-]+|[\s,-]+$/g, '');
}

/** Same rules as festrec_eval/genres.py map_genres: whole phrase, then head word. */
export function mapGenres(phrases) {
  const found = new Set();
  for (const phrase of phrases) {
    for (let part of cleanGenre(phrase).split(/\s*[/,·|]\s*/)) {
      part = part.trim();
      if (!part) continue;
      if (COMPOUNDS[part]) { COMPOUNDS[part].forEach((g) => found.add(g)); continue; }
      if (WORDS[part]) { found.add(WORDS[part]); continue; }
      const words = part.split(/\s+/);
      const head = words[words.length - 1];
      if (WORDS[head]) { found.add(WORDS[head]); continue; }
      const pieces = head.split('-').filter((p) => WORDS[p]);
      if (pieces.length) { pieces.forEach((p) => found.add(WORDS[p])); continue; }
      words.filter((w) => WORDS[w]).forEach((w) => found.add(WORDS[w]));
    }
  }
  return found;
}

async function api(url, params) {
  const query = new URLSearchParams({ ...params, format: 'json', origin: '*' });
  const response = await fetch(`${url}?${query}`);
  if (!response.ok) throw new Error(`Wikidata answered ${response.status}`);
  return response.json();
}

const label = (entity) =>
  entity?.labels?.en?.value || entity?.labels?.mul?.value ||
  Object.values(entity?.labels || {})[0]?.value || '';

function claimIds(entity, prop) {
  const statements = entity?.claims?.[prop] || [];
  const preferred = statements.filter((s) => s.rank === 'preferred');
  const ordered = preferred.length ? preferred : statements.filter((s) => s.rank !== 'deprecated');
  const ids = [];
  for (const statement of ordered) {
    const id = statement.mainsnak?.datavalue?.value?.id;
    if (id && !ids.includes(id)) ids.push(id);
  }
  return ids;
}

function releaseYear(entity) {
  const years = (entity?.claims?.P577 || [])
    .map((s) => String(s.mainsnak?.datavalue?.value?.time || '').match(/[+-]?(\d{4})/)?.[1])
    .filter(Boolean);
  return years.length ? years.sort()[0] : '';
}

function runtimeMinutes(entity) {
  for (const statement of entity?.claims?.P2047 || []) {
    const value = statement.mainsnak?.datavalue?.value;
    const amount = Number(value?.amount);
    if (!Number.isFinite(amount)) continue;
    const unit = String(value.unit || '').split('/').pop();
    const minutes = unit === 'Q25235' ? amount * 60 : unit === 'Q11574' ? amount / 60 : amount;
    if (minutes >= 1 && minutes <= 1000) return Math.round(minutes);
  }
  return null;
}

async function entities(ids, props) {
  const out = {};
  for (let i = 0; i < ids.length; i += 50) {
    const payload = await api(WIKIDATA, {
      action: 'wbgetentities', ids: ids.slice(i, i + 50).join('|'), props,
      languages: 'en|mul', languagefallback: '1', sitefilter: 'enwiki',
    });
    Object.assign(out, payload.entities || {});
  }
  return out;
}

function trimOverview(text) {
  const clean = String(text || '').replace(/\s+/g, ' ').trim();
  if (clean.length <= OVERVIEW_CHARS) return clean;
  const cut = clean.slice(0, OVERVIEW_CHARS);
  const end = Math.max(cut.lastIndexOf('. '), cut.lastIndexOf('! '), cut.lastIndexOf('? '));
  return end > 200 ? cut.slice(0, end + 1) : `${cut.slice(0, cut.lastIndexOf(' '))}…`;
}

export class WikidataProvider {
  constructor(cache) {
    this.name = 'wikidata';
    this.cache = cache;
  }

  async lookup(title, year) {
    const cacheKey = `wd:${String(title).toLowerCase()}|${year || ''}`;
    const cached = await this.cache?.get(cacheKey);
    if (cached !== undefined) return cached;
    const found = await this.fetchFilm(title, year);
    await this.cache?.set(cacheKey, found);
    return found;
  }

  /** Search by title, keep only films, and prefer the one from the right year. */
  async fetchFilm(title, year) {
    const search = await api(WIKIDATA, {
      action: 'wbsearchentities', search: title, language: 'en', type: 'item', limit: '10',
    });
    const ids = (search.search || []).map((item) => item.id);
    if (!ids.length) return null;

    const candidates = Object.values(await entities(ids, 'labels|claims|sitelinks'))
      .filter((entity) => claimIds(entity, 'P31').some((type) => FILM_TYPES.has(type)));
    if (!candidates.length) return null;

    const distance = (entity) => {
      const found = Number(releaseYear(entity));
      return year && found ? Math.abs(found - Number(year)) : 5;
    };
    candidates.sort((a, b) => distance(a) - distance(b));
    const film = candidates[0];
    // Two films can share a title; a year that's far off means it isn't this one.
    if (year && distance(film) > 1) return null;

    const props = ['P57', 'P58', 'P161', 'P725', 'P136', 'P921'];
    const wanted = [...new Set(props.flatMap((p) => {
      const found = claimIds(film, p);
      return p === 'P161' || p === 'P725' ? found.slice(0, CAST_DEPTH) : found;
    }))];
    const people = wanted.length ? await entities(wanted, 'labels') : {};
    const named = (prop, limit) =>
      claimIds(film, prop).slice(0, limit).map((id) => label(people[id])).filter(Boolean);

    const genreWords = named('P136');
    const genres = mapGenres(genreWords);
    claimIds(film, 'P31').forEach((type) => TYPE_GENRES[type] && genres.add(TYPE_GENRES[type]));
    const keywords = [
      ...named('P921').map((k) => k.toLowerCase()),
      ...genreWords.map(cleanGenre).filter((g) => g && !GENRES.includes(g)),
    ];

    const article = film.sitelinks?.enwiki?.title || '';
    let overview = '';
    if (article) {
      const page = await api(WIKIPEDIA, {
        action: 'query', prop: 'extracts', exintro: '1', explaintext: '1',
        redirects: '1', titles: article, formatversion: '2',
      });
      overview = trimOverview(page.query?.pages?.[0]?.extract);
    }

    return {
      title: label(film),
      year: releaseYear(film),
      runtime: runtimeMinutes(film),
      director: named('P57'),
      writer: named('P58'),
      cast: named('P161', CAST_DEPTH).length ? named('P161', CAST_DEPTH) : named('P725', CAST_DEPTH),
      keyword: [...new Set(keywords)],
      genre: [...genres].sort(),
      overview,
      wikipedia: article,
    };
  }
}

File: festrec_eval/wikidata.py
"""Film credits from Wikidata, synopses from Wikipedia.

Wikidata's data is CC0 - free for any use, including fitting a model and
shipping what it knows inside an app. Wikipedia's text is CC BY-SA 4.0, so any
synopsis taken from it keeps its article title for attribution.

Both services ask to be used politely: a descriptive User-Agent, batched
requests, and backing off when told to. Everything fetched is cached, so a
build only ever asks once per film.

Records come out in the shape the features expect:
    {title, year, runtime, genre, director, writer, cast, keyword, country,
     language, overview, wikipedia, qid, imdb}
"""

from __future__ import annotations

import json
import re
import time
import urllib.error
import urllib.parse
import urllib.request
from concurrent.futures import ThreadPoolExecutor
from pathlib import Path

from .genres import GENRES, clean, map_genres

USER_AGENT = (
    "festival-schedule-builder/0.2 "
    "(https://github.com/mostmetaphorical/festival-schedule-builder)"
)
WIKIDATA_API = "https://www.wikidata.org/w/api.php"
SPARQL = "https://query.wikidata.org/sparql"
WIKIPEDIA_API = "https://en.wikipedia.org/w/api.php"

CAST_DEPTH = 6  # beyond the top billing, shared actors stop meaning much
OVERVIEW_CHARS = 700
BATCH = 50  # wbgetentities' limit per request
EXTRACT_BATCH = 20  # prop=extracts' limit per request

# Things that are films. Instance-of paths through subclasses are too slow to
# query for a whole year, so the common film types are listed directly.
FILM_TYPES = (
    "Q11424",     # film
    "Q24869",     # feature film
    "Q29168811",  # animated feature film
    "Q202866",    # animated film
    "Q93204",     # documentary film
    "Q226730",    # silent film
    "Q506240",    # television film
)

MINUTE, HOUR, SECOND = "Q7727", "Q25235", "Q11574"


def request_json(url: str, params: dict, attempts: int = 6) -> dict:
    """GET with a proper User-Agent, retrying on rate limits and lag."""
    query = urllib.parse.urlencode(params)
    req = urllib.request.Request(
        f"{url}?{query}",
        headers={"User-Agent": USER_AGENT, "Accept": "application/json"},
    )
    for attempt in range(attempts):
        try:
            with urllib.request.urlopen(req, timeout=90) as response:
                payload = json.loads(response.read().decode("utf-8"))
            # The API reports replication lag as an error with a retry hint.
            if isinstance(payload, dict) and payload.get("error", {}).get("code") == "maxlag":
                time.sleep(5 + 2**attempt)
                continue
            return payload
        except urllib.error.HTTPError as error:
            if error.code in (429, 500, 502, 503, 504):
                time.sleep(float(error.headers.get("Retry-After") or 2**attempt + 1))
                continue
            raise
        except (urllib.error.URLError, TimeoutError, json.JSONDecodeError):
            time.sleep(2**attempt + 1)
    raise RuntimeError(f"gave up on {url} after {attempts} attempts")


def sparql(query: str) -> list[dict]:
    payload = request_json(SPARQL, {"query": query, "format": "json"})
    return payload["results"]["bindings"]


def chunks(items: list, size: int):
    for start in range(0, len(items), size):
        yield items[start:start + size]


def qids_for_imdb(imdb_ids: list[str]) -> dict[str, str]:
    """IMDb title IDs -> Wikidata items. Wikidata records the IMDb ID of most films."""
    found: dict[str, str] = {}
    for batch in chunks(sorted(set(imdb_ids)), 300):
        values = " ".join(f'"{i}"' for i in batch)
        for row in sparql(f"SELECT ?imdb ?film WHERE {{ VALUES ?imdb {{ {values} }} ?film wdt:P345 ?imdb . }}"):
            found.setdefault(row["imdb"]["value"], row["film"]["value"].rsplit("/", 1)[1])
    return found


def popular_films(year: int, limit: int) -> list[str]:
    """The films of a year that the most Wikipedias write about.

    Sitelinks are a decent proxy for "widely seen" that needs no ratings data.
    """
    types = " ".join(f"wd:{t}" for t in FILM_TYPES)
    rows = sparql(f"""
        SELECT ?film (MAX(?links) AS ?n) WHERE {{
          VALUES ?type {{ {types} }}
          ?film wdt:P31 ?type ; wdt:P577 ?date ; wikibase:sitelinks ?links .
          FILTER(?date >= "{year}-01-01"^^xsd:dateTime && ?date < "{year + 1}-01-01"^^xsd:dateTime)
          FILTER(?links >= 3)
        }} GROUP BY ?film ORDER BY DESC(?n) LIMIT {limit}
    """)
    return [row["film"]["value"].rsplit("/", 1)[1] for row in rows]


def get_entities(ids: list[str], props: str, workers: int = 4) -> dict[str, dict]:
    """wbgetentities in batches of 50, a few at a time."""
    def fetch(batch: list[str]) -> dict:
        payload = request_json(WIKIDATA_API, {
            "action": "wbgetentities", "ids": "|".join(batch), "props": props,
            "languages": "en|mul", "languagefallback": "1", "sitefilter": "enwiki",
            "format": "json", "maxlag": "5",
        })
        return payload.get("entities", {})

    entities: dict[str, dict] = {}
    with ThreadPoolExecutor(max_workers=workers) as pool:
        for result in pool.map(fetch, list(chunks(sorted(set(ids)), BATCH))):
            entities.update({k: v for k, v in result.items() if "missing" not in v})
    return entities


def label(entity: dict) -> str:
    labels = entity.get("labels", {})
    for language in ("en", "mul"):
        if language in labels:
            return labels[language]["value"]
    return next(iter(labels.values()), {}).get("value", "")


def claim_ids(entity: dict, prop: str) -> list[str]:
    """Item values of a property, in the order they're listed, best rank first."""
    statements = entity.get("claims", {}).get(prop, [])
    preferred = [s for s in statements if s.get("rank") == "preferred"]
    ordered = preferred or [s for s in statements if s.get("rank") != "deprecated"]
    ids = []
    for statement in ordered:
        value = statement.get("mainsnak", {}).get("datavalue", {}).get("value")
        if isinstance(value, dict) and value.get("id") and value["id"] not in ids:
            ids.append(value["id"])
    return ids


def claim_strings(entity: dict, prop: str) -> list[str]:
    values = []
    for statement in entity.get("claims", {}).get(prop, []):
        value = statement.get("mainsnak", {}).get("datavalue", {}).get("value")
        if isinstance(value, str):
            values.append(value)
    return values


def release_year(entity: dict) -> str:
    years = []
    for statement in entity.get("claims", {}).get("P577", []):
        value = statement.get("mainsnak", {}).get("datavalue", {}).get("value", {})
        match = re.match(r"[+-]?(\d{4})", str(value.get("time", "")))
        if match:
            years.append(match.group(1))
    return min(years) if years else ""


def runtime_minutes(entity: dict) -> int | None:
    for statement in entity.get("claims", {}).get("P2047", []):
        value = statement.get("mainsnak", {}).get("datavalue", {}).get("value", {})
        try:
            amount = float(value.get("amount"))
        except (TypeError, ValueError):
            continue
        unit = str(value.get("unit", "")).rsplit("/", 1)[-1]
        minutes = amount * 60 if unit == HOUR else amount / 60 if unit == SECOND else amount
        if 1 <= minutes <= 1000:
            return int(round(minutes))
    return None


def people_and_things(films: dict[str, dict]) -> list[str]:
    """Every item a film record points at, so their names can be fetched once."""
    wanted = set()
    for entity in films.values():
        for prop in ("P57", "P58", "P161", "P725", "P136", "P921", "P495", "P364"):
            ids = claim_ids(entity, prop)
            wanted.update(ids[:CAST_DEPTH] if prop in ("P161", "P725") else ids)
    return sorted(wanted)


# Film types that say something genre does not: Wikidata files "animated" and
# "documentary" as what a film is, and often leaves them out of its genres.
TYPE_GENRES = {
    "Q29168811": "animation", "Q202866": "animation", "Q17517379": "animation",
    "Q93204": "documentary", "Q24865": "documentary",
}


def extracts(titles: list[str], workers: int = 3) -> dict[str, str]:
    """Plain-text lead paragraphs of English Wikipedia articles."""
    def fetch(batch: list[str]) -> dict[str, str]:
        payload = request_json(WIKIPEDIA_API, {
            "action": "query", "prop": "extracts", "exintro": "1", "explaintext": "1",
            "redirects": "1", "titles": "|".join(batch), "format": "json",
            "formatversion": "2", "maxlag": "5",
        })
        query = payload.get("query", {})
        # Map back through normalisation and redirects to the asked-for title.
        back = {}
        for step in query.get("normalized", []) + query.get("redirects", []):
            back[step["to"]] = back.get(step["from"], step["from"])
        found = {}
        for page in query.get("pages", []):
            if page.get("extract"):
                asked = back.get(page["title"], page["title"])
                found[asked] = page["extract"]
        return found

    out: dict[str, str] = {}
    with ThreadPoolExecutor(max_workers=workers) as pool:
        for result in pool.map(fetch, list(chunks(sorted(set(titles)), EXTRACT_BATCH))):
            out.update(result)
    return out


def trim_overview(text: str) -> str:
    """The lead's first few sentences, which is where a synopsis lives."""
    text = re.sub(r"\s+", " ", text or "").strip()
    if len(text) <= OVERVIEW_CHARS:
        return text
    cut = text[:OVERVIEW_CHARS]
    end = max(cut.rfind(". "), cut.rfind("! "), cut.rfind("? "))
    return cut[: end + 1] if end > 200 else cut.rsplit(" ", 1)[0] + "…"


def genres_for(words: list[str], types: list[str]) -> list[str]:
    genres = map_genres(words)
    genres.update(TYPE_GENRES[t] for t in types if t in TYPE_GENRES)
    return sorted(genres)


def keywords_for(subjects: list[str], words: list[str]) -> list[str]:
    # A specific genre ("folk horror") is too fine for the genre list but is
    # exactly what a keyword is for.
    fine = [clean(g) for g in words]
    return list(dict.fromkeys(
        [s.lower() for s in subjects] + [f for f in fine if f and f not in GENRES]
    ))


def condense(entity: dict, names: dict[str, str], extract: str = "") -> dict:
    """One film, in the shape the features and the app expect."""
    def named(prop: str, limit: int | None = None) -> list[str]:
        ids = claim_ids(entity, prop)
        ids = ids[:limit] if limit else ids
        return [names[i] for i in ids if names.get(i)]

    genre_words = named("P136")
    subjects = named("P921")
    article = entity.get("sitelinks", {}).get("enwiki", {}).get("title", "")
    title = label(entity) or re.sub(r"\s*\((\d{4} )?film\)$", "", article)

    types = [t for t in claim_ids(entity, "P31") if t in TYPE_GENRES]

    return {
        "title": title,
        "year": release_year(entity),
        "runtime": runtime_minutes(entity),
        "genre": genres_for(genre_words, types),
        # Kept so a change to the genre mapping can be re-applied without
        # fetching every film again (see FilmCache.remap).
        "genre_source": {"words": genre_words, "types": types},
        "director": named("P57"),
        "writer": named("P58"),
        # An animated film's performers are its voice cast.
        "cast": named("P161", CAST_DEPTH) or named("P725", CAST_DEPTH),
        "keyword": keywords_for(subjects, genre_words),
        "keyword_source": subjects,
        "country": named("P495"),
        "language": named("P364"),
        "overview": trim_overview(extract),
        "wikipedia": article,
        "qid": entity.get("id", ""),
        "imdb": next(iter(claim_strings(entity, "P345")), ""),
    }


class FilmCache:
    """Condensed film records by Wikidata ID, kept on disk between runs."""

    def __init__(self, path: str | Path):
        self.path = Path(path)
        self.records: dict[str, dict] = (
            json.loads(self.path.read_text(encoding="utf-8")) if self.path.exists() else {}
        )

    def remap(self) -> None:
        """Re-apply the current genre and keyword rules to cached films."""
        for record in self.records.values():
            source = (record or {}).get("genre_source")
            if source is None:
                continue
            record["genre"] = genres_for(source["words"], source["types"])
            record["keyword"] = keywords_for(record.get("keyword_source", []), source["words"])

    def save(self) -> None:
        self.path.parent.mkdir(parents=True, exist_ok=True)
        self.path.write_text(json.dumps(self.records, ensure_ascii=False), encoding="utf-8")

    def fill_overviews(self, qids: list[str], step: int = 1000, log=print) -> None:
        """Add Wikipedia synopses to cached films fetched without them.

        Resumable: films already holding a synopsis are skipped, and progress
        is saved after each step. A film whose article has no extract is marked
        with an empty string so it isn't asked for again.
        """
        todo = [
            q for q in dict.fromkeys(qids)
            if (record := self.records.get(q))
            and record.get("wikipedia")
            and not record.get("overview")
            and not record.get("overview_checked")
        ]
        log(f"fetching synopses for {len(todo)} films")
        for start in range(0, len(todo), step):
            batch = todo[start:start + step]
            texts = extracts([self.records[q]["wikipedia"] for q in batch])
            for qid in batch:
                record = self.records[qid]
                record["overview"] = trim_overview(texts.get(record["wikipedia"], ""))
                record["overview_checked"] = True
            self.save()
            log(f"  {min(start + step, len(todo))}/{len(todo)}")

    def fill(self, qids: list[str], overviews: bool = True, step: int = 2000,
             log=print) -> None:
        """Fetch whatever isn't cached yet, saving after each step."""
        # Records from before genre sources were kept can't be remapped, so
        # they are fetched again.
        todo = [
            q for q in dict.fromkeys(qids)
            if q not in self.records
            or (self.records[q] is not None and "genre_source" not in self.records[q])
        ]
        log(f"{len(self.records)} films cached, fetching {len(todo)}")
        for start in range(0, len(todo), step):
            batch = todo[start:start + step]
            films = get_entities(batch, "labels|claims|sitelinks")
            names = {
                qid: label(entity)
                for qid, entity in get_entities(people_and_things(films), "labels").items()
            }
            articles = {
                qid: entity.get("sitelinks", {}).get("enwiki", {}).get("title")
                for qid, entity in films.items()
            }
            texts = extracts([a for a in articles.values() if a]) if overviews else {}
            for qid, entity in films.items():
                self.records[qid] = condense(entity, names, texts.get(articles[qid] or "", ""))
            # Not found at all: remember that too, so it isn't asked again.
            for qid in batch:
                self.records.setdefault(qid, None)
            self.save()
            log(f"  {min(start + step, len(todo))}/{len(todo)}")
